pkg/controllers/provisioning/provisioner.go

```diff
@@ -17,122 +17,142 @@ package provisioning
+	"github.com/awslabs/karpenter/pkg/utils/pretty"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/prometheus/client_golang/prometheus"
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/workqueue"
 	"knative.dev/pkg/logging"
-)
-
-var (
-	MaxBatchDuration = time.Second * 10
-	MinBatchDuration = time.Second * 1
-	// MaxPodsPerBatch limits the number of pods we process at one time to avoid using too much memory
-	MaxPodsPerBatch = 2_000
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
 )
 
 // Provisioner waits for enqueued pods, batches them, creates capacity and binds the pods to the capacity.
 type Provisioner struct {
 	// State
 	*v1alpha5.Provisioner
-	pods    chan *v1.Pod
-	results chan error
-	done    <-chan struct{}
+	Batches sync.Map
+	Ctx     context.Context
 	Stop    context.CancelFunc
 	// Dependencies
 	cloudProvider cloudprovider.CloudProvider
-	scheduler     *scheduling.Scheduler
-	launcher      *Launcher
+	coreV1Client  corev1.CoreV1Interface
+	packer        *binpacking.Packer
 }
 
-func (p *Provisioner) Start(ctx context.Context) {
-	go func() {
-		logging.FromContext(ctx).Info("Starting provisioner")
-		for {
-			select {
-			case <-p.done:
-				logging.FromContext(ctx).Info("Stopping provisioner")
-				return
-			default:
-				if err := p.provision(ctx); err != nil {
-					logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
-				}
-			}
-		}
-	}()
-}
-
-func (p *Provisioner) provision(ctx context.Context) (err error) {
-	// Wait for a batch of pods
-	pods := p.Batch(ctx)
-	// Communicate the result of the provisioning loop to each of the pods.
-	defer func() {
-		for i := 0; i < len(pods); i++ {
-			select {
-			case p.results <- err: // Block until result is communicated
-			case <-p.done: // Leave if closed
-			}
-		}
-	}()
-	// Separate pods by scheduling constraints
-	schedules, err := p.scheduler.Solve(ctx, p.Provisioner, pods)
+// Add (threadsafe) the pod to the provisioner and block until it's processed.
+// The caller is responsible for verifying that the pod was scheduled correctly.
+func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod, constraints *v1alpha5.Constraints) {
+	key, err := hashstructure.Hash(constraints, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
 	if err != nil {
-		return fmt.Errorf("solving scheduling constraints, %w", err)
+		logging.FromContext(ctx).Fatalf("Failed add pod to provisioner, hashing constraints, %s, %w", pretty.Concise(constraints), err.Error())
 	}
-	// Get Instance Types, offering availability may vary over time
-	instanceTypes, err := p.cloudProvider.GetInstanceTypes(ctx, &p.Spec.Constraints)
-	if err != nil {
-		return fmt.Errorf("getting instance types")
+	batch, ok := p.Batches.LoadOrStore(key, NewBatch(ctx, fmt.Sprint(key), constraints))
```

Comment:
It could just be me, but I think this code is a bit confusing ... since `key` is derived from `constraints`, the fact that it's repeated twice (passed to both `NewBatch` and `LoadOrStore`) comes across as unusual. It feels like since the `Batch` struct already knows about `Constraints`, it should maybe deal with this problem (of not starting more than one provisioner per batch?) rather than force the caller to do so?